### PR TITLE
Fix: TypeScript Type Portability Issues

### PR DESCRIPTION
**Pull request by aryaemami59**
# **Overview**

The type portability issues in this repository are primarily caused by three specific TypeScript errors:

1. [**`TS4023`**](https://github.com/microsoft/TypeScript/issues/37888): `Exported variable 'useAppDispatch' has or is using name 'TimeResponse' from external module "/home/runner/work/redux-toolkit/redux-toolkit/examples/publish-ci/vite/src/app/services/times" but cannot be named.`

   - **Solution**: This one can actually be resolved quite often by converting the problematic type from an `interface` to a `type` alias. Other times we might have to export the problematic type.

2. [**`TS2742`**](https://typescript.tv/errors/#ts2742): `The inferred type of 'useGetQuotesQuery' cannot be named without a reference to '../../../node_modules/@reduxjs/toolkit/dist/query/react/buildHooks'. This is likely not portable. A type annotation is necessary.`

   - **Solution**: Bundling the type definitions seems to have resolved this issue.

3. `TS2527`: `The inferred type of 'quotesApiSlice' references an inaccessible 'unique symbol' type. A type annotation is necessary.`
   This one took a while for me to figure out, It turns out in order for a `unique symbol` type to be "accessible", the `unique symbol` has to be imported upon declaration, so we want something that looks like this:

   ```ts
   export declare const coreModuleName: unique symbol
   ```

   But `dts-rollup-plugin` gives us this instead:

   ```ts
   declare const coreModuleName: unique symbol
   export { coreModuleName }
   ```

   But TypeScript treats `declare const coreModuleName: unique symbol` and `export { coreModuleName };` as 2 different `unique symbol`s because each reference to `coreModuleName` is completely unique.

   - **Solution**: While somewhat hacky, I wrote a small script that takes all exported `unique symbol`s and converts them from:

     ```ts
     declare const coreModuleName: unique symbol
     export { coreModuleName }
     ```

     To:

     ```ts
     export declare const coreModuleName: unique symbol
     ```

## **This PR:**

- [x] Resolves:

  - #1806
  - #2314
  - #2485
  - #3202
  - #3568
  - #3962
  - #3970
  - #3983
  - #4066
  - #4108
  - #4401
  - #4448
  - #4492
  - #4498
  - #4518

- [x] Converts the following types from `interface` to `type` alias:

  - `AsyncThunkSliceReducerConfig`
  - `ReducerDefinition`
  - `AsyncThunkSliceReducerDefinition`
  - `MutationTypes`
  - `MutationThunkArg`
  - `MutationCacheLifecycleApi`
  - `MutationLifecycleApi`
  - `BaseEndpointTypes`
  - `CacheLifecyclePromises`
  - `MutationBaseLifecycleApi`
  - `QueryLifecyclePromises`
  - `LifecycleApi`
  - `PromiseWithKnownReason`
  - `QueryTypes`
  - `StartQueryActionCreatorOptions`
  - `QueryThunkArg`
  - `UseQuerySubscriptionOptions`
  - `GetSelectorsOptions`
  - `EndpointDefinitionWithQueryFn`
  - `EndpointDefinitionWithQuery`
  - `BaseActionCreator`
  - `WithMiddleware`
  - `ReactDynamicMiddlewareInstance`
  - `AddListenerOverloads`
  - `TypedActionCreatorWithMatchFunction`
  - `TypedActionCreator`
  - `CaseReducerDefinition`
  - `InjectIntoConfig`
  - `InjectedSlice`
  - `QueryHooks`
  - `MutationHooks`
  - `ThunkWithReturnValue`

- [x] Exports the following types:

  - `StartQueryActionCreatorOptions`
  - `CombinedSliceReducer`
  - `BaseEndpointDefinition`
  - `AddMiddleware`
  - `CreateDispatchWithMiddlewareHook`
  - `DynamicDispatch`

- [x] Exports the following values:

  - `UNINITIALIZED_VALUE`
  - `_NEVER`

- [x] Removes relative `declare module` statements.
- [x] Fixes issues related to name collision of types. Sometimes `tsup` has a hard time with aliasing type imports so if we have something like this:

  ```ts
  import type { Dispatch as ReduxDispatch } from 'redux'
  ```

  `tsup` somehow renames `ReduxDispatch` back to `Dispatch` which causes issues when you have a generic type parameter with the same name so this:

  ```ts
  export type AddMiddleware<
    State = any,
    Dispatch extends ReduxDispatch<UnknownAction> = ReduxDispatch<UnknownAction>
  > = {
    (...middlewares: Middleware<any, State, Dispatch>[]): void
    withTypes<MiddlewareConfig extends MiddlewareApiConfig>(): AddMiddleware<
      GetState<MiddlewareConfig>,
      GetDispatch<MiddlewareConfig>
    >
  }
  ```

  Becomes:

  ```ts
  export type AddMiddleware<
    State = any,
    Dispatch extends Dispatch<UnknownAction> = Dispatch<UnknownAction>
  > = {
    (...middlewares: Middleware<any, State, Dispatch>[]): void
    withTypes<MiddlewareConfig extends MiddlewareApiConfig>(): AddMiddleware<
      GetState<MiddlewareConfig>,
      GetDispatch<MiddlewareConfig>
    >
  }
  ```

  Which can cause issues as well as make things confusing. This issue is fixed by doing 2 things:

  1. Renaming some generic type parameters to prevent name collision with imported external types:

     ```diff
     export type AddMiddleware<
       State = any,
     - Dispatch extends ReduxDispatch<UnknownAction> = ReduxDispatch<UnknownAction>,
     + DispatchType extends ReduxDispatch<UnknownAction> = ReduxDispatch<UnknownAction>,
       > = {
      (...middlewares: Middleware<any, State, DispatchType>[]): void
      withTypes<MiddlewareConfig extends MiddlewareApiConfig>(): AddMiddleware<
       GetState<MiddlewareConfig>,
           GetDispatch<MiddlewareConfig>
         >
      }
     ```

  2. Removing some import aliases:

     ```diff
     - import type { Dispatch as ReduxDispatch } from 'redux'
     + import type { Dispatch } from 'redux'
     ```

- [x] Patches `console-testing-library` so we can run the type tests with `"moduleResolution": "Bundler"`. Previously it would fail due to `console-testing-library` not having a `types` field in it's subpath exports.
- [x] Adds 3 new example workspaces:

1. **`@examples-type-portability/bundler`** to test for `"moduleResolution": "Bundler"`.
2. **`@examples-type-portability/nodenext-cjs`** to test for `"moduleResolution": "NodeNext"` with TypeScript's CJS syntax and `"type": "commonjs"` in `package.json`.
3. **`@examples-type-portability/nodenext-esm`** to test for `"moduleResolution": "NodeNext"` with ESM syntax and `"type": "module"` in `package.json`.

## **Things to keep in mind**

- Although I prefer using `interfaces`, it might be worthwhile to switch to `type` aliases going forward. `interfaces` are more prone to type portability issues compared to `type` aliases, which often dissolve into their constituents. I highly recommend using `type` aliases for all new types unless `interfaces` are necessary (e.g., for declaration merging).

- I've noticed that using an `interface` as the return type of an external function often leads to type portability issues unless the `interface` itself is also exported.

### CODE REVIEW

**Comment by EskiMojo14** [comment]
this is going to cause a painful amount of conflicts with other PRs 😅

**Comment by aryaemami59** [comment]
I know, I'm sorry :( but I really think this will be worth it.

**Comment by aryaemami59** [comment]
Just out of curiosity what are some of the other PRs that are gonna have conflicts with this one? And is there anything I can do to minimize said conflicts?

**Comment by EskiMojo14** [comment]
the changes in this PR are very wide reaching (including some renames that seem unrelated to the issue at hand?), so i suspect quite a few will encounter issues, but in particular the creator PR #3837 is quite significant in how much it reworks, and will certainly encounter issues when trying to reconcile with this.

**Comment by aryaemami59** [comment]
The renames were necessary, they were causing issues with name collisions, I'll see if I can work around it to minimize conflicts with other PRs.

**Comment by aryaemami59** [comment]
Alright this one should be good to go. Let me know if this needs anything.

File: packages/toolkit/src/createAsyncThunk.ts [hunk 1]
```diff
@@ -125,15 +118,10 @@ export type AsyncThunkConfig = {
...
-type GetState<ThunkApiConfig> = ThunkApiConfig extends {
-  state: infer State
-}
-  ? State
-  : unknown
 type GetExtra<ThunkApiConfig> = ThunkApiConfig extends { extra: infer Extra }
   ? Extra
   : unknown
-type GetDispatch<ThunkApiConfig> = ThunkApiConfig extends {
+type GetDispatchType<ThunkApiConfig> = ThunkApiConfig extends {
```
Comment (markerikson) [inline packages/toolkit/src/createAsyncThunk.ts:124]: :question: what required the renames?

File: packages/toolkit/src/dynamicMiddleware/index.ts [hunk 2]
```diff
@@ -1,27 +1,29 @@
...
+  DynamicMiddlewareInstance,
+  GetDispatch,
+  GetState,
+  MiddlewareApiConfig,
 } from './types'
 
 const createMiddlewareEntry = <
   State = any,
-  Dispatch extends ReduxDispatch<UnknownAction> = ReduxDispatch<UnknownAction>,
+  DispatchType extends Dispatch<UnknownAction> = Dispatch<UnknownAction>,
```
Comment (markerikson) [inline packages/toolkit/src/dynamicMiddleware/index.ts:23]: :question: This seems cosmetic - why was this needed?

File: packages/toolkit/src/query/apiTypes.ts [hunk 3]
```diff
@@ -1,30 +1,358 @@
...
   BaseQuery extends BaseQueryFn,
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   Definitions extends EndpointDefinitions,
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   ReducerPath extends string,
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   TagTypes extends string,
-> {}
+> {
+  [coreModuleName]: {
```
Comment (markerikson) [inline packages/toolkit/src/query/apiTypes.ts:53]: :question: I assume this was added because we're no longer doing interface merging, or something along those lines?

**Review by markerikson — APPROVED** [review]
Probably have more questions later, but these are the first things that come to mind.

File: packages/toolkit/src/createAsyncThunk.ts [hunk 4]
```diff
@@ -1,24 +1,17 @@
 import type { Dispatch, UnknownAction } from 'redux'
-import type {
-  PayloadAction,
-  ActionCreatorWithPreparedPayload,
-} from './createAction'
-import { createAction } from './createAction'
 import type { ThunkDispatch } from 'redux-thunk'
+import type { ActionCreatorWithPreparedPayload } from './createAction'
+import { createAction } from './createAction'
+import type { GetState } from './dynamicMiddleware/types'
```
[the remark this answers is not in the document]
Comment (aryaemami59) [reply packages/toolkit/src/createAsyncThunk.ts:5]: I did it to de-duplicate the `GetState` type to prevent having `GetState$1` in the type declaration file.

File: packages/toolkit/src/createAsyncThunk.ts (hunk 1 above)
Comment (aryaemami59) [reply packages/toolkit/src/createAsyncThunk.ts:124]: This one was done to avoid name collision with `GetDispatch` type in `src/dynamicMiddleware/types.ts`.

File: packages/toolkit/src/dynamicMiddleware/index.ts (hunk 2 above)
Comment (aryaemami59) [reply packages/toolkit/src/dynamicMiddleware/index.ts:23]: The renames were done because `tsup` sometimes has a hard time with aliasing type imports so something like:

```ts
import type { Dispatch as ReduxDispatch } from 'redux'
```

It somehow renames `ReduxDispatch` back to `Dispatch` which causes issues when you have a generic type parameter with the same name:

```ts
export type AddMiddleware<
  State = any,
  Dispatch extends ReduxDispatch<UnknownAction> = ReduxDispatch<UnknownAction>,
> = {
  (...middlewares: Middleware<any, State, Dispatch>[]): void
  withTypes<MiddlewareConfig extends MiddlewareApiConfig>(): AddMiddleware<
    GetState<MiddlewareConfig>,
    GetDispatch<MiddlewareConfig>
  >
}
```

which can turn into:

```ts
export type AddMiddleware<
  State = any,
  Dispatch extends Dispatch<UnknownAction> = Dispatch<UnknownAction>,
> = {
  (...middlewares: Middleware<any, State, Dispatch>[]): void
  withTypes<MiddlewareConfig extends MiddlewareApiConfig>(): AddMiddleware<
    GetState<MiddlewareConfig>,
    GetDispatch<MiddlewareConfig>
  >
}
```

Which can cause issues.

File: packages/toolkit/src/query/apiTypes.ts (hunk 3 above)
Comment (aryaemami59) [reply packages/toolkit/src/query/apiTypes.ts:53]: We can still do `interface` merging, I only removed `declare module`s that were using relative paths. Since now we are bundling the type defs, and relative paths will no longer be applicable.

File: packages/toolkit/src/createAsyncThunk.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (markerikson) [reply packages/toolkit/src/createAsyncThunk.ts:5]: Could we either rename the type from `dynamicMiddleware` instead, or consolidate into one type somehow?

File: packages/toolkit/src/createAsyncThunk.ts (hunk 1 above)
Comment (markerikson) [reply packages/toolkit/src/createAsyncThunk.ts:124]: Same thing. I'd rather keep the `cAT` version as the "primary" - seems like there'd be less churn overall, and `dynamicMiddleware` is both a newer file and less important.

Comment (aryaemami59) [reply packages/toolkit/src/createAsyncThunk.ts:124]: Sure thing

Comment (aryaemami59) [reply packages/toolkit/src/createAsyncThunk.ts:129]: The only issue is the `GetDispatch` type from `src/dynamicMiddleware/types.ts` is the one that is exported through the main entry point, so if I rename that one, it'll be a breaking change.

**Comment by markerikson** [comment]
Okay, changes here make sense based on the descriptions, and I think Arya's got it cleaned up to where this is basically "just" the `interface -> type` changes (rather than moving type definitions or contents between files).

Let's get this in.
